_posts/2021-02-11-tabnet/tabnet.Rmd

```diff
@@ -0,0 +1,411 @@
+    $ m_wbb                    <dbl> 0.9887509, 0.9923558, 0.8659244, 1…
+    $ m_wwbb                   <dbl> 0.8766783, 0.7983426, 0.7801176, 0…
+
+11,000,000 "observations" (kind of) -- that's a lot! Like the authors of the TabNet paper (@arik2020tabnet), we'll use 500,000 of these for validation. (Unlike them, though, we won't be able to train for 870,000 iterations!)
+
+The first variable, `class`, is either `1` or `0`, depending on whether a Higgs boson was present or not. While in experiments, only a tiny fraction of collisions produce one of those, both classes are about equally frequent in this dataset.
+
+As to the predictors, the last seven are high-level (derived). All others are "measured".
+
+Data loaded, we're ready to build a `tidymodels` *workflow*, resulting in a short sequence of concise steps.
+
+First, split the data:
+
+```{r}
+n <- 11000000
+n_test <- 500000
+test_frac <- n/n_all
+
+split <- initial_time_split(higgs, prop = 1 - test_frac)
+train <- training(split)
+test  <- testing(split)
+```
+
+Second, create a `recipe`. We want to predict `class` from all other features present:
+
+```{r}
+rec <- recipe(class ~ ., train) 
+```
+
+Third, create a `parsnip` model specification of class `tabnet`. The parameters passed are those reported, by the TabNet paper, for the S-sized model variant used on this dataset. [^2]
+
+[^2]: Apart from the number of epochs, that is.
+
+```{r}
+# hyperparameter settings (apart from epochs) as per the TabNet paper (TabNet-S)
+mod <- tabnet(epochs = 3, batch_size = 16384, decision_width = 24, attention_width = 26,
+              num_steps = 5, penalty = 0.000001, virtual_batch_size = 512, momentum = 0.6,
+              feature_reusage = 1.5, learn_rate = 0.02) %>%
+  set_engine("torch", verbose = TRUE) %>%
+  set_mode("classification")
+
+```
+
+Fourth, bundle recipe and model specification in a workflow:
+
+```{r}
+wf <- workflow() %>%
+  add_model(mod) %>%
+  add_recipe(rec)
+```
+
+Fifth, train the model. This will take some time. Training finished, we save the trained `parsnip` model, so we can reuse it at a later time.
+
+```{r}
+fitted_model <- wf %>% fit(train)
+
+# access the underlying parsnip model and save it to RDS format
+# depending on when you read this, a nice wrapper may exist
+# see https://github.com/mlverse/tabnet/issues/27  
+fitted_model$fit$fit$fit %>% saveRDS("saved_model.rds")
+```
+
+After three epochs, loss was at 0.609.
+
+Sixth - and finally -- we ask the model for test set predictions and have accuracy computed.
+
+```{r}
+preds <- test %>% 
+  bind_cols(predict(fitted_model, test))
+
+yardstick::accuracy(preds, class, .pred_class)
+```
+
+    # A tibble: 1 x 3
+      .metric  .estimator .estimate
+      <chr>    <chr>          <dbl>
+    1 accuracy binary         0.672
+
+We didn't quite arrive at the accuracy reported in the TabNet paper (0.783), but then, we trained for a tiny fraction of the time.
```

Comment:
```suggestion
We didn't quite arrive at the accuracy reported in the TabNet paper (0.783), but then, we only trained for a tiny fraction of the time.
```

Reply:
😜  giving you back one of the _only_s I took away elsewhere!

Reply:
lol!!!